Updated BOM Generator

Xelynega opened the pull request:
Added part references and whether or not a part is no_load. Fixed bug with part numbers. Changed dependency from python elementTree to lxml because elemenTree cannot move up/down parent/child elements in xml files.

Xelynega committed:
Updated BOM Generator

Added part references and whether or not a part is no_load. Fixed bug with part numbers. Changed dependency from python elementTree to lxml because elemenTree cannot move up/down parent/child elements in xml files.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,65 @@
+        #Check if the element is under a component or a library element
+        if field.getparent().getparent().tag == 'comp':
+            no_load = False
+            #Check if the part is a no_load part
```
It would be preferable to add information about all the available fields instead of adding specifically the no_load field. This way the script won't need maintenance if we need to add a parameter to the components.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,65 @@
+
+try:
+    from lxml import etree as ET
+except ImportError:
```
I've never seen this kind of import before. looks a lot cleaner for installation, I like it.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,65 @@
+        if "part_num" not in line:
+            print "line doesn't have part_num! line: ", line
+        elif line['part_num'] in found_parts:
+            output_dict[found_parts.index(line['part_num'])].update(line)
```
If there is no component in the component_library.csv file, there should be some kind of warning printed out to the user. I should have added this error check initially but forgot to

scholvat reviewed (changes requested):
Looks good, definately an improvement. It would be better if the script could fetch all of the fields for a component instead of specifically getting the no_load field.

Xelynega committed:
Added new features

Added ability for new fields to be added in KiCad without any need for refactoring the script. Added warning message for parts that are in the netlist, but missing from the component library.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,104 @@
+
+try:
+    from lxml import etree as ET
+except ImportError:
```
Why did you change from xml to lxml? xml comes standard in python, lxml needs to be installed

scholvat reviewed (commented):
Can you explain why you switched xml libraries, or switch back to the one that is included in python?

Xelynega committed:
Removed 3rd party dependencies

Changed library to python xml

Xelynega commented:
I used lxml because it's the only one I've used before and because I had no idea how to use XPath. Changed it to use pythons built-in library and found out that XPath isn't that bad.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,101 @@
+            else:
+                output_dict[found_parts.index(part)]['quantity'] += 1
+                output_dict[found_parts.index(part)]['ref'] += (', ' + ref)
+    reader = csv.DictReader(open('component_library.csv', 'rb'))
```
component_library.csv is not in the same directory as the python script. See the original code I wrote to get the proper filename, and either do something similar to (or exactly) that, or move the python script into the same root directory.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,101 @@
+
+    #Get a list of all the keys for the CSV at the end
+    for part_element in tree.findall(".//*[@name='part_num']...."):
+        print part_element
```
Remove debug print statement

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,101 @@
+
+    #Give error with missing parts
+    warning_message = "\n\tWARNING: Parts "
+    for p in missing_parts: 
```
Don't use p as a variable name

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,101 @@
+    missing_parts = []
+    for part in output_dict:
+        try:   
+            if part['Distributer #'] != '':
```
Parts can be in the WARG library and not have distributer part numbers. Maybe add some logic to line 74 to check if there is a matching part number in the component library.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,101 @@
+            missing_parts.append(part)
+
+    #Give error with missing parts
+    warning_message = "\n\tWARNING: Parts "
```
You can have have a separate print statement for each warning, there isn't a need for this loop, it just complicates the code.

scholvat reviewed (changes requested):
I tried running the program for the first time and found some more bugs

Xelynega committed:
Fixed bugs

Changed the variable named 'p', changed missing part error from looking for distributor numbers to looking for part_name entries in the component library, removed debug lines, changed path for component library to look in the directory above the script in the hierarchy.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,92 @@
+            else:
+                output_dict[found_parts.index(part)]['quantity'] += 1
+                output_dict[found_parts.index(part)]['ref'] += (', ' + ref)
+    reader = csv.DictReader(open('../component_library.csv', 'rb'))
```
Still not working on my machine... maybe us the os library like I originally had.

scholvat commented on bom/generate_warg_bom.py:
```diff
@@ -1,49 +1,92 @@
-
-if __name__ == "__main__":
-    main()
+import pip
```
remove pip since I don't think you need it anymore

Xelynega committed:
Changed file paths

Changed from using text-only file path to using the os library for opening the component library.

Xelynega committed:
Removed pip library

scholvat commented:
Looks good, I tested it on both Window and Linux and runs fine.
